source/docs/software/commandbased/organizing-command-based.rst

```diff
@@ -0,0 +1,259 @@
+    )
+
+  .. code-tab:: c++
+
+    // TODO
+
+(Creating one ``StartEndCommand`` instance and putting it in a variable won't work here, since once an instance of a command is added to a command group it is effectively "owned" by that command group and cannot be used in any other context.)
+
+Subclassing CommandBase
+^^^^^^^^^^^^^^^^^^^^^^^
+
+One possible solution to this problem is to create a new subclass of ``CommandBase`` that implements the necessary ``initialize`` and ``end`` routines.
+
+.. tabs::
+
+  .. code-tab:: java
+
+    public class RunIntakeCommand extends CommandBase {
+        private Intake m_intake;
+
+        public RunIntakeCommand(Intake intake) {
+            this.m_intake = intake;
+            addRequirements(intake);
+        }
+
+        @Override
+        public void initialize() {
+            m_intake.setPercent(1);
+        }
+
+        @Override
+        public void end(boolean interrupted) {
+            m_intake.setPercent(0);
+        }
+
+        @Override
+        public boolean isFinished() {
+            return false;
+        }
+    }
+
+  .. code-tab:: c++
+
+    // TODO
+
+This, however, is just as cumbersome. The only two lines that really matter in this entire file are the two calls to ``intake.setPercent``, yet there are over 20 lines of boilerplate code! Not to mention, doing this for a lot of robot actions quickly clutters up a robot project with dozens of small files. Nevertheless, this might feel more "natural," particularly for programmers who don't feel confident with Java and want to stick closely to an object-oriented model.
+
+Factory Methods
+^^^^^^^^^^^^^^^
+
+A useful middle ground between these two extremes is using a `factory method <https://www.tutorialspoint.com/design_pattern/factory_pattern.htm>`__. A factory method is a method that, each time it is called, returns a new object according to some specification.
+
+A command like the intake-running command is conceptually related to exactly one subsystem: the ``Intake``. As such, it makes sense to put a ``commandRun`` method as an instance method of the ``Intake`` class:
+
+.. tabs::
+
+  .. code-tab:: java
+
+    public class Intake {
+        // [code for motor controllers, configuration, etc.]
+        // ...
+
+        public Command commandRun() {
+            return new StartEndCommand(() -> this.set(1), () -> this.set(0), this);
+        }
+    }
+
+  .. code-tab:: c++
+
+    // TODO
+
+Notice how since we are in the ``Intake`` class, we no longer refer to ``intake``; instead, we use the ``this`` keyword to refer to the current instance.
+
+.. warning:: To preserve encapsulation, avoid referring to ``private`` variables and methods of the ``Intake`` class from within the ``commandRun`` method, even though they are technically accessible. (For instance, the ``commandRun`` method should use the subsystem's public ``set`` method instead of interfacing directly with the motor controllers.)
```

Comment:
I disagree with this. If already preserving encapsulation, I'd advise *not* having public `set` methods and exposing only command factories (and getters, when they are needed from outside the subsystem). See [6738's 2022 code for example](https://github.com/ExcaliburFRC/RapidReact2022/blob/main/src/main/java/io/excaliburfrc/robot/subsystems/Intake.java).
It also mostly solves the namespace collision issue.

Reply:
I think it would be better in the instance factory method to use the motor object directly, and then add a comment somewhere that in some cases it might be better to extract logic to non-command methods (and then teams would make them public if needed).

Reply:
I strongly dislike giving any `Command` code direct knowledge of its subsystem's hardware setup (consider the scenario where a motor controller gets changed, or motors are added or removed).

I think I'm going to try to have another discussion of subsystem encapsulation here as well (it's looking like maybe this should be split into multiple documents), so maybe we have a link here to that discussion but for now we should just take a position and stick to it while we talk about other things (namely where commands live). Sort of the same deal as the enum thing.

Reply:
Swapping motor controllers shouldn't change anything, that's what polymorphism exists for. Motors being added/removed would either need to modify commands anyway or it wouldn't at all.
And encapsulation isn't a problem here; nothing escapes outside.
What caused my stance on this is seeing three levels of indirection with an `arcadeDriveCommand` factory method that called an `arcadeDrive` method that called `DifferentialDrive.arcadeDrive()`: the middle level is more than redundant.